hack/generator/pkg/codegen/storage/storage_type_factory.go

```diff
@@ -0,0 +1,465 @@
+	types                      astmodel.Types             // All the types for this service
+	propertyConversions        []propertyConversion       // Conversion rules to use for properties when creating storage variants
+	pendingStorageConversion   astmodel.TypeNameQueue     // Queue of types that need storage variants created for them
+	pendingConversionInjection astmodel.TypeNameQueue     // Queue of types that need conversion functions injected
+	pendingMarkAsHubVersion    astmodel.TypeNameQueue     // Queue of types that need to be flagged as the hub storage version
+	idFactory                  astmodel.IdentifierFactory // Factory for creating identifiers
+	storageConverter           astmodel.TypeVisitor       // a cached type visitor used to create storage variants
+	propertyConverter          astmodel.TypeVisitor       // a cached type visitor used to simplify property types
+	functionInjector           astmodel.TypeVisitor       // a cached type visitor used to inject functions into definitions
+	resourceHubMarker          astmodel.TypeVisitor       // a cached type visitor used to mark resources as Storage Versions
+
+	// Map of conversion links for creating our conversion graph
+	// (Can't use PackageReferences as keys, so keyed by the full package path)
+	conversionMap map[string]astmodel.PackageReference
+}
+
+// NewStorageTypeFactory creates a new instance of StorageTypeFactory ready for use
+func NewStorageTypeFactory(service string, idFactory astmodel.IdentifierFactory) *StorageTypeFactory {
+	result := &StorageTypeFactory{
+		service:                    service,
+		types:                      make(astmodel.Types),
+		pendingStorageConversion:   astmodel.MakeTypeNameQueue(),
+		pendingConversionInjection: astmodel.MakeTypeNameQueue(),
+		pendingMarkAsHubVersion:    astmodel.MakeTypeNameQueue(),
+		idFactory:                  idFactory,
+		conversionMap:              make(map[string]astmodel.PackageReference),
+	}
+
+	result.propertyConversions = []propertyConversion{
+		result.preserveKubernetesResourceStorageProperties,
+		result.defaultPropertyConversion,
+	}
+
+	result.storageConverter = astmodel.TypeVisitorBuilder{
+		VisitObjectType:    result.convertObjectType,
+		VisitResourceType:  result.convertResourceType,
+		VisitTypeName:      result.redirectTypeNamesToStoragePackage,
+		VisitValidatedType: result.stripAllValidations,
+		VisitFlaggedType:   result.stripAllFlags,
+	}.Build()
+
+	result.propertyConverter = astmodel.TypeVisitorBuilder{
+		VisitEnumType:      result.useBaseTypeForEnumerations,
+		VisitValidatedType: result.stripAllValidations,
+		VisitTypeName:      result.shortCircuitNamesOfSimpleTypes,
+	}.Build()
+
+	result.functionInjector = astmodel.TypeVisitorBuilder{
+		VisitObjectType:   result.injectFunctionIntoObject,
+		VisitResourceType: result.injectFunctionIntoResource,
+	}.Build()
+
+	result.resourceHubMarker = astmodel.TypeVisitorBuilder{
+		VisitResourceType: result.markResourceAsStorageVersion,
+	}.Build()
```

Comment:
I think these 4 type visitors should be separated out of `StorageTypeFactory` - probably as 4 separate types each with their own callback methods. I haven't read through them yet but I'm assuming the callbacks aren't all manipulating shared state on result? Splitting them out into separate types would make that clearer (or at least make it obvious what state is shared between them), and hopefully make each visitor more coherent - as it is they're all intermingled into one type.